airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java

```diff
@@ -0,0 +1,218 @@
+
+@Slf4j
+public class ProtocolVersionChecker {
+
+  private final JobPersistence jobPersistence;
+  private final Configs configs;
+  private final ConfigRepository configRepository;
+  private final DefinitionsProvider definitionsProvider;
+
+  // Dependencies could be simplified once we break some pieces up:
+  // * JobPersistence for accessing the airbyte_metadata table.
+  // * Configs for getting the new Airbyte Protocol Range from the env vars.
+  // * ConfigRepository for accessing ActorDefinitions
+  public ProtocolVersionChecker(final JobPersistence jobPersistence,
+                                final Configs configs,
+                                final ConfigRepository configRepository,
+                                final DefinitionsProvider definitionsProvider) {
+    this.jobPersistence = jobPersistence;
+    this.configs = configs;
+    this.configRepository = configRepository;
+    this.definitionsProvider = definitionsProvider;
+  }
+
+  /**
+   * Validate the AirbyteProtocolVersion support range between the platform and the connectors.
+   * <p>
+   * The goal is to make sure that we do not end up disabling existing connections after an upgrade
+   * that changes the protocol support range.
+   *
+   * @param supportAutoUpgrade whether the connectors will be automatically upgraded by the platform
+   * @return the supported protocol version range if check is successful, Optional.empty() if we would
+   *         break existing connections.
+   * @throws IOException
+   */
+  public Optional<AirbyteProtocolVersionRange> validate(final boolean supportAutoUpgrade) throws IOException {
+    final Optional<AirbyteVersion> currentAirbyteVersion = getCurrentAirbyteVersion();
+    final AirbyteProtocolVersionRange currentRange = getCurrentProtocolVersionRange();
+    final AirbyteProtocolVersionRange targetRange = getTargetProtocolVersionRange();
+
+    // Checking if there is a pre-existing version of airbyte.
+    // Without this check, the first run of the validation would fail because we do not have the tables
+    // set yet
+    // which means that the actor definitions lookup will throw SQLExceptions.
+    if (currentAirbyteVersion.isEmpty()) {
+      log.info("No previous version of Airbyte detected, assuming this is a fresh deploy.");
+      return Optional.of(targetRange);
+    }
+
+    if (currentRange.equals(targetRange)) {
+      log.info("Using AirbyteProtocolVersion range [{}:{}]", targetRange.min().serialize(), targetRange.max().serialize());
+      return Optional.of(targetRange);
+    }
+
+    log.info("Detected an AirbyteProtocolVersion range change from [{}:{}] to [{}:{}]",
+        currentRange.min().serialize(), currentRange.max().serialize(),
+        targetRange.min().serialize(), targetRange.max().serialize());
+
+    final Map<ActorType, Set<UUID>> conflicts = getConflictingActorDefinitions(targetRange);
+
+    if (conflicts.isEmpty()) {
+      log.info("No conflicts");
```

Comment:
Nit:  Consider adding more context to this message so it's easy to understand what has no conflicts.  Perhaps something like "No conflicting connector protocol versions detected."

Reply:
updated